src/keys/bip39/mod.rs

```diff
@@ -0,0 +1,750 @@
+        let descriptor_key = self
+            .into_extended_key()?
+            .into_descriptor_key(source, derivation_path)?;
+
+        Ok(set_valid_on_any_network(descriptor_key))
+    }
+}
+
+#[cfg_attr(docsrs, doc(cfg(feature = "keys-bip39")))]
+impl<Ctx: ScriptContext> GeneratableKey<Ctx> for Mnemonic {
+    type Entropy = Box<[u8]>;
+
+    // If word count is `None`, it is picked from the entropy length
+    type Options = (Option<WordCount>, Language);
+    type Error = Option<Error>;
+
+    fn generate_with_entropy(
+        (word_count, language): Self::Options, entropy: Self::Entropy
+    ) -> Result<GeneratedKey<Self, Ctx>, Self::Error> {
+        let entropy = match word_count {
+            Some(count) => &entropy.as_ref()[..(count as usize / 8)],
+            None => &entropy
+        };
+        let mnemonic = Mnemonic::from_entropy_in(language, entropy)?;
+
+        Ok(GeneratedKey::new(mnemonic, any_network()))
+    }
+
+    fn generate(options: Self::Options) -> Result<GeneratedKey<Self, Ctx>, Self::Error> {
+        use rand::{thread_rng, Rng};
+
+        let mut entropy = Box::new([0; 32]);
+        thread_rng().fill(entropy.as_mut());
+        Self::generate_with_entropy(options, entropy)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::str::FromStr;
+    use super::{Language, Mnemonic};
+    use bitcoin::{util::bip32, hashes::hex::FromHex};
+    use crate::keys::{any_network, GeneratableKey, GeneratedKey, bip39::WordCount};
+
+    #[test]
+    fn test_keys_bip39_mnemonic() {
+        let mnemonic =
+            "aim bunker wash balance finish force paper analyst cabin spoon stable organ";
+        let mnemonic = Mnemonic::parse_in(Language::English, mnemonic).unwrap();
+        let path = bip32::DerivationPath::from_str("m/44'/0'/0'/0").unwrap();
+
+        let key = (mnemonic, path);
+        let (desc, keys, networks) = crate::descriptor!(wpkh(key)).unwrap();
+        assert_eq!(desc.to_string(), "wpkh([be83839f/44'/0'/0']xpub6DCQ1YcqvZtSwGWMrwHELPehjWV3f2MGZ69yBADTxFEUAoLwb5Mp5GniQK6tTp3AgbngVz9zEFbBJUPVnkG7LFYt8QMTfbrNqs6FNEwAPKA/0/*)#0r8v4nkv");
+        assert_eq!(keys.len(), 1);
+        assert_eq!(networks.len(), 4);
+    }
+
+    #[test]
+    fn test_keys_bip39_mnemonic_passphrase() {
+        let mnemonic =
+            "aim bunker wash balance finish force paper analyst cabin spoon stable organ";
+        let mnemonic = Mnemonic::parse_in(Language::English, mnemonic).unwrap();
+        let path = bip32::DerivationPath::from_str("m/44'/0'/0'/0").unwrap();
+
+        let key = ((mnemonic, Some("passphrase".into())), path);
+        let (desc, keys, networks) = crate::descriptor!(wpkh(key)).unwrap();
+        assert_eq!(desc.to_string(), "wpkh([8f6cb80c/44'/0'/0']xpub6DWYS8bbihFevy29M4cbw4ZR3P5E12jB8R88gBDWCTCNpYiDHhYWNywrCF9VZQYagzPmsZpxXpytzSoxynyeFr4ZyzheVjnpLKuse4fiwZw/0/*)#h0j0tg5m");
+        assert_eq!(keys.len(), 1);
+        assert_eq!(networks.len(), 4);
+    }
+
```

Comment:
You might want to also add test for invalid mnemonic sentences.

```suggestion
//less than 12 words
assert_eq!(parse_in(Language::English, "advice cage absurd amount doctor acoustic avoid letter advice cage above").unwrap(), Err(Error::InvalidWordCount(11)));

//more than 24 words
assert_eq!(parse_in(Language::English, "abandon ability able about above absent absorb abstract absurd abuse access accident account accuse achieve acid acoustic acquire across act action actor actress actual adapt add addict").unwrap(), Err(Error::InvalidWordCount(27)));

//not a multiple of six 
assert_eq!(parse_in(Language::English, "abandon ability able about above absent absorb abstract absurd abuse access accident account accuse").unwrap(), Err(Error::InvalidWordCount(14)));

//invalid word
assert_eq!(parse_in(Language::English, "getter advice cage absurd amount doctor acoustic avoid letter advice cage above").unwrap(), Err(Error::InvalidWord(String::from("getter")));

//Once you implement the verification of the checksum in `parse_in` you might want to test that below here:
```

Reply:
While adding tests for incorrect entropies do you want to add these tests too?

Reply:
I'm not sure I understand you properly. Are you asking if it is necessary to test for these given that we are already testing incorrect entropies? If yes, then I think it is. I never saw a test for `parse_in` and I think it is necessary to have it.

Reply:
I think these tests should be there. I meant since you've written the code already, probably you should commit it as well.

Reply:
@atalw , I have added verification and also the tests I suggested above.

Reply:
@atalw , I have added verification and also the tests I suggested above.

Reply:
@atalw , I have added verification and also the tests I suggested above.